### DISCUSSION TITLE
Mode 1 and Mode 2 from MCP23xx or PCF8574

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#22523] [Q&A]
I ask it on other thread ends but i am still not sure.

Both modes of both modules does support more than one module, doesn't?

The modes differs only in the behavior like system perepherie (button,relay,...) or does not (with different options and output)?

**Comment by sfromis** [comment]
As the docs explains, MCP230xx mode 1 is **one** chip, while mode 2 can be multiple
https://tasmota.github.io/docs/MCP230xx/#supporting-modes

And the two modes are *very* different in use, but it is not clear what your questions aims to clarify. Or why you ask about mode 1 vs mode 2 again and again.

**Comment by ChrisVonFrosta** [comment]
The reason is that I built the entire button press logic (and PWM lights) in Berry script, since mode 2 wasn't introduced for IO expanders at that time. I was so confused about mode 2 from those IO expanders because my first encounter with a "mode 2" was through using the PCA9685. This was necessary to get the option to use more than one PCA9685. Ok, in the meanwhile i checked that PCA9685 has no mode2 but a V2 driver:  PCA9685_V2.

So, before I developed my entire logic for buttons using mode 1 for one PCF8574, which works fine.

Actually, I switched to two MCP23017. Since I realized that I need mode 2 here again (like V2 from PCA9685), I concluded that this is the main behavior of mode 2. But as you've mentioned many times: This is not the biggest difference.

I inquired again about using mode 1 with multiple MCP23017 modules, to be able to use most of my existing script and button processing (single, double, ..., long push). Now, I started with using Tasmota's built-in button engine (Mode 2), which makes my script smaller and maybe the button processing a bit faster. However, I already noticed that using "SINGLE" and "DOUBLE" instead of "TOGGLE" introduces a delay. Or does that just mean delay needed to be able to detect a subsequent button press? Well, that goes without saying.

I still need to test the mentions about `SwitchMode 11/12`. But i think that will be bring all functions needed.
thx!

**Reply by sfromis** [reply]
When it comes to the inherently necessary delay for detecting `Button` multi-press (instead of single), `SetOption13` exists for "immediate action", disabling multi-press.

While I've not tested anything with MCP23017, I'd expect this normal button processing behavior to apply here too, as a main point of mode 2 is to be able to treat the expander pins "just like" regular gpios (as long as it is Button/Switch/Relay/Output).
